04_documentation/code_documentation_tools_slides.md

```diff
@@ -0,0 +1,363 @@
+---
+type: slide
+slideOptions:
+  transition: slide
+  width: 1400
+  height: 900
+  margin: 0.1
+---
+
+<style>
+  .reveal strong {
+    font-weight: bold;
+    color: orange;
+  }
+  .reveal p {
+    text-align: left;
+  }
+  .reveal section h1 {
+    color: orange;
+  }
+  .reveal section h2 {
+    color: orange;
+  }
+  .reveal code {
+    font-family: 'Ubuntu Mono';
+    color: orange;
+  }
+  .reveal section img {
+    background:none;
+    border:none;
+    box-shadow:none;
+  }
+</style>
+
+# Code Documentation Tools
+
+---
+
+## Learning Goals
+
+- You know about common ways to generate code documentation.
+- You know tools for Python and C++.
+- You know common website generators for documentation.
+- You know how to publish your homepage.
+
+---
+
+## Introduction
+
+- Software projects need documentation
+- Different types of documentation and locations
+    - Code (functions, API etc.)
+    - Everything else (tutorials, project description, contribution guides...)
+- [Docs as Code](https://www.writethedocs.org/guide/docs-as-code/), [DocOps](https://www.writethedocs.org/guide/doc-ops/)
+
+---
+
+## Documentation (Tool) Requirements
+
+- Documentation should be easy to write
+    - Lower hurdle to actually document software
+- Tools should enhance presentation of documentation
+    - Formatting, searching...
+- Documentation should be accessible
+    - Host website
+
+---
+
+## Short Tool Overview
+
+- General documentation website generators
+    - [MkDocs](https://www.mkdocs.org) (md), [Sphinx](https://www.sphinx-doc.org/en/master/) (rst),...
+- Code documentation website generators
+    - [Doxygen](https://www.doxygen.nl) (C++, C...), Sphinx with [autodoc extension](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) (Python),...
+- Publishing tools
+    - [Read the Docs](https://readthedocs.org/), [GitHub Pages](https://pages.github.com/)...
+- Example: [SSE homepage](https://simulation-software-engineering.github.io/homepage/)
+    - MkDocs + [MkDocs material theme](https://squidfunk.github.io/mkdocs-material/) + GitHub Pages
+
+---
+
+## Step by Step Plan
+
+1. Document Python code
+2. Create (general) documentation website
+3. Add code documentation to this websites
+4. Publish the homepage via Read the Docs
+
+The code is on [GitHub](https://github.com/Simulation-Software-Engineering/python-code-documentation)
+
+---
+
+## Docstrings
+
+```python
+"""Docstring explaining the script/module"""
+
+class ClassName:
+    """Docstring explaining the class"""
+
+    def awesome_function(self, parameter ):
+        """A docstring explaining the function"""
+        return parameter
+```
+
+- [PEP 257 - Docstring Conventions](https://www.python.org/dev/peps/pep-0257/)
+- `__doc__` property is derived from docstring
+- May span multiple lines
```

Comment:
... and first line has special meaning (short description), see "technical writing" part.

Reply:
This comes later/is only common for multi-line docstrings and special docstring formatting standards like the ones from Google or Sphinx. I wanted to mention this here, but not enforce it yet. It will be more obvious/important when we reformat the docstrings.